llvm/tools/sycl-post-link/sycl-post-link.cpp

```diff
@@ -274,21 +273,56 @@ static void collectKernelModuleMap(
 
 // Go through function call graph searching for assert call.
 static bool hasAssertInFunctionCallGraph(llvm::Function *Func) {
+  // Map holds the info about assertions in already examined functions:
+  // true  - if there is an assertion in underlying functions,
+  // false - if there are definetely no assertions in underlying functions.
+  static std::map<llvm::Function *, bool> hasAssertionInCallGraphMap;
+  std::vector<llvm::Function *> FuncList;
+
   std::vector<llvm::Function *> Workqueue;
   Workqueue.push_back(Func);
 
   while (!Workqueue.empty()) {
     Function *F = Workqueue.back();
     Workqueue.pop_back();
+    if (F != Func)
+      FuncList.push_back(F);
+
+    bool IsVertex = true;
     for (auto &I : instructions(F)) {
       if (CallBase *CB = dyn_cast<CallBase>(&I))
         if (Function *CF = CB->getCalledFunction()) {
-          if (CF->getName().startswith("__devicelib_assert_fail"))
+          // Return if we've already discovered if there are asserts in the
+          // function call graph.
+          if (hasAssertionInCallGraphMap.count(CF)) {
+            return hasAssertionInCallGraphMap[CF];
+          }
+
+          if (CF->getName().startswith("__devicelib_assert_fail")) {
+            // Mark all the functions above in call graph as ones that can call
+            // assert.
+            for (auto *It : FuncList)
+              hasAssertionInCallGraphMap[It] = true;
```

Comment:
This doesn't seem to be correct, i.e. `FuncList` does not represent the current call stack, consider the following example:
```
A -> B
A -> C
A -> D
D -> assert
```

While doing BFS, we would enter `A`, then push `B`, `C` and `D` into queue. Both `B` and `C` do not use `assert`, but they will be added into `FuncList` and therefore, once you reach `D`, you will mark `B` and `C` as using assertions, which is incorrect.

If you want to maintain the current call stack, then you should probably switch to using DFS for graph traversal.

However, you can still achieve the result you want with BFS and additional data structure to store backedges/parents/callers: if you have a map `callee` -> `list of callers`, then at any point you can traverse the graph upwards marking each node as using `assert`